File: web/src/lib/masker.test.ts
import { detectPIIType, isLuhnValid, maskValue, maskRecord } from './masker.ts'

declare const process: { exit: (code: number) => void }

let passed = 0
let failed = 0

function assert(condition: boolean, msg: string) {
  if (!condition) {
    throw new Error(`Assertion failed: ${msg}`)
  }
}

function test(name: string, fn: () => void) {
  try {
    fn()
    passed++
    console.log(`  ✓ ${name}`)
  } catch (err: any) {
    failed++
    console.error(`  ✗ ${name}: ${err.message}`)
  }
}

console.log('--- Running Masker & PII Detection Unit Tests ---')

test('isLuhnValid accepts valid card numbers', () => {
  assert(isLuhnValid('4532015112830366'), 'Visa 16 digits')
  assert(isLuhnValid('4532 0151 1283 0366'), 'Visa with spaces')
  assert(isLuhnValid('378282246310005'), 'Amex 15 digits')
})

test('isLuhnValid rejects invalid card numbers', () => {
  assert(!isLuhnValid('4532015112830367'), 'Checksum invalid')
  assert(!isLuhnValid('1234'), 'Too short')
  assert(!isLuhnValid(''), 'Empty')
})

test('detectPIIType detects PII by column and value', () => {
  assert(detectPIIType('email') === 'email', 'email col')
  assert(detectPIIType('notes', 'alice@domain.com') === 'email', 'email val')
  assert(detectPIIType('phone') === 'phone', 'phone col')
  assert(detectPIIType('contact', '+1 (555) 234-5678') === 'phone', 'phone val')
  assert(detectPIIType('credit_card') === 'card', 'card col')
  assert(detectPIIType('raw', '4532015112830366') === 'card', 'card val')
  assert(detectPIIType('ssn') === 'ssn', 'ssn col')
  assert(detectPIIType('id_num', '123-45-6789') === 'ssn', 'ssn val')
  assert(detectPIIType('ip_address') === 'ip', 'ip col')
  assert(detectPIIType('host', '192.168.1.1') === 'ip', 'ip val')
  assert(detectPIIType('first_name') === 'name', 'name col')
  assert(detectPIIType('table_name') === '', 'technical table_name not pii')
  assert(detectPIIType('id', '123') === '', 'id not pii')
})

test('maskValue: redact strategy', () => {
  const res = maskValue('email', 'alice@domain.com', 'redact')
  assert(res === '[REDACTED]', `expected [REDACTED], got ${res}`)
})

test('maskValue: partial strategy', () => {
  const email = maskValue('email', 'john.doe@example.com', 'partial')
  assert(email.startsWith('j***@') && email.endsWith('@example.com'), `partial email: ${email}`)

  const card = maskValue('card', '4532015112830366', 'partial')
  assert(card.includes('••••') && card.endsWith('0366'), `partial card: ${card}`)

  const ssn = maskValue('ssn', '123-45-6789', 'partial')
  assert(ssn === '•••-••-6789', `partial ssn: ${ssn}`)

  const pass = maskValue('password', 'secret', 'partial')
  assert(pass === '••••••••', `partial pass: ${pass}`)
})

test('maskValue: hash strategy', () => {
  const h1 = maskValue('email', 'alice@domain.com', 'hash')
  assert(typeof h1 === 'string' && h1.startsWith('hash_'), `hash email: ${h1}`)
  const h2 = maskValue('email', 'alice@domain.com', 'hash')
  assert(h1 === h2, 'deterministic hash')
})

test('maskValue: faker strategy', () => {
  const f1 = maskValue('email', 'alice@domain.com', 'faker')
  assert(typeof f1 === 'string' && f1.includes('@'), `faker email: ${f1}`)
  const f2 = maskValue('email', 'alice@domain.com', 'faker')
  assert(f1 === f2, 'deterministic faker stability')

  const card = maskValue('card', '4532015112830366', 'faker')
  assert(isLuhnValid(card), `faker card passes Luhn check: ${card}`)
})

test('maskRecord masks only PII columns', () => {
  const cols = ['id', 'email', 'name', 'created_at']
  const record = ['42', 'test@example.com', 'Bob Smith', '2026-01-01']
  const masked = maskRecord(cols, record, 'redact')
  assert(masked[0] === '42', 'id untouched')
  assert(masked[1] === '[REDACTED]', 'email redacted')
  assert(masked[2] === '[REDACTED]', 'name redacted')
  assert(masked[3] === '2026-01-01', 'created_at untouched')
})

console.log(`\nMasker Tests: ${passed} passed, ${failed} failed`)
if (failed > 0) process.exit(1)

File: internal/masker/masker_test.go
package masker

import (
	"reflect"
	"strings"
	"testing"
)

func TestIsLuhnValid(t *testing.T) {
	// Valid Luhn numbers
	validCards := []string{
		"4532015112830366",
		"4532 0151 1283 0366",
		"4532-0151-1283-0366",
		"378282246310005",
	}
	for _, c := range validCards {
		if !IsLuhnValid(c) {
			t.Errorf("expected Luhn valid for %s, got false", c)
		}
	}

	// Invalid Luhn numbers
	invalidCards := []string{
		"4532015112830367",
		"1234",
		"abcdefghij123",
		"",
	}
	for _, c := range invalidCards {
		if IsLuhnValid(c) {
			t.Errorf("expected Luhn invalid for %s, got true", c)
		}
	}
}

func TestDetectPIIType(t *testing.T) {
	tests := []struct {
		colName  string
		sample   string
		expected string
	}{
		{"email", "", PIITypeEmail},
		{"user_email", "foo@bar.com", PIITypeEmail},
		{"notes", "contact@test.org", PIITypeEmail},
		{"phone", "", PIITypePhone},
		{"mobile_no", "+1 (555) 123-4567", PIITypePhone},
		{"untyped", "+12345678901", PIITypePhone},
		{"credit_card", "", PIITypeCard},
		{"custom", "4532015112830366", PIITypeCard},
		{"ssn", "", PIITypeSSN},
		{"national_id", "", PIITypeSSN},
		{"gov", "123-45-6789", PIITypeSSN},
		{"ip_address", "", PIITypeIP},
		{"host", "192.168.1.10", PIITypeIP},
		{"ipv6_node", "2001:0db8:85a3:0000:0000:8a2e:0370:7334", PIITypeIP},
		{"password", "my_secret", PIITypePassword},
		{"api_key", "secret-token-123", PIITypePassword},
		{"first_name", "Alice", PIITypeName},
		{"last_name", "Smith", PIITypeName},
		{"name", "John Doe", PIITypeName},
		{"full_name", "Jane Doe", PIITypeName},
		{"street", "123 Main St", PIITypeAddress},
		{"city", "Springfield", PIITypeAddress},
		// Non-PII cases
		{"id", "123", ""},
		{"table_name", "users", ""},
		{"column_name", "email", ""},
		{"filename", "photo.png", ""},
		{"created_at", "2026-01-01T00:00:00Z", ""},
		{"status", "active", ""},
		{"price", "99.99", ""},
		{"notes", strings.Repeat("a", 257) + "@example.com", ""},
	}

	for _, tt := range tests {
		got := DetectPIIType(tt.colName, tt.sample)
		if got != tt.expected {
			t.Errorf("DetectPIIType(%q, %q) = %q, expected %q", tt.colName, tt.sample, got, tt.expected)
		}
	}
}

func TestMaskValueStrategies(t *testing.T) {
	// 1. Redaction
	r1 := MaskValue("email", "alice@example.com", StrategyRedact)
	if r1 != "[REDACTED]" {
		t.Errorf("StrategyRedact failed, got: %v", r1)
	}

	// 2. Hash
	h1 := MaskValue("email", "alice@example.com", StrategyHash)
	hStr, ok := h1.(string)
	if !ok || !strings.HasPrefix(hStr, "hash_") {
		t.Errorf("StrategyHash failed, got: %v", h1)
	}

	// 3. Partial
	pEmail := MaskValue("email", "john.doe@example.com", StrategyPartial).(string)
	if !strings.HasPrefix(pEmail, "j***@") || !strings.HasSuffix(pEmail, "@example.com") {
		t.Errorf("StrategyPartial email unexpected: %s", pEmail)
	}

	pCard := MaskValue("card", "4532015112830366", StrategyPartial).(string)
	if !strings.HasSuffix(pCard, "0366") || !strings.Contains(pCard, "••••") {
		t.Errorf("StrategyPartial card unexpected: %s", pCard)
	}

	pPhone := MaskValue("phone", "+1 555 123 4567", StrategyPartial).(string)
	if !strings.HasSuffix(pPhone, "4567") || !strings.Contains(pPhone, "•••") {
		t.Errorf("StrategyPartial phone unexpected: %s", pPhone)
	}

	pSSN := MaskValue("ssn", "123-45-6789", StrategyPartial).(string)
	if pSSN != "•••-••-6789" {
		t.Errorf("StrategyPartial ssn unexpected: %s", pSSN)
	}

	pName := MaskValue("name", "Alice Wonderland", StrategyPartial).(string)
	if !strings.Contains(pName, "A***e") || !strings.Contains(pName, "W***d") {
		t.Errorf("StrategyPartial name unexpected: %s", pName)
	}

	pPass := MaskValue("password", "supersecret123", StrategyPartial).(string)
	if pPass != "••••••••" {
		t.Errorf("StrategyPartial password unexpected: %s", pPass)
	}

	// 4. Faker
	fName := MaskValue("name", "John Smith", StrategyFaker).(string)
	if fName == "John Smith" || !strings.Contains(fName, " ") {
		t.Errorf("StrategyFaker name unexpected: %s", fName)
	}

	fEmail := MaskValue("email", "john.doe@example.com", StrategyFaker).(string)
	if !strings.Contains(fEmail, "@") {
		t.Errorf("StrategyFaker email unexpected: %s", fEmail)
	}

	fCard := MaskValue("card", "4532015112830366", StrategyFaker).(string)
	if !IsLuhnValid(fCard) {
		t.Errorf("StrategyFaker card is not valid Luhn: %s", fCard)
	}
}

func TestDeterministicFakerStability(t *testing.T) {
	inputVal := "corporate.vip@bigcorp.io"
	first := MaskValue("email", inputVal, StrategyFaker)

	for i := 0; i < 20; i++ {
		repeat := MaskValue("email", inputVal, StrategyFaker)
		if repeat != first {
			t.Fatalf("Faker is not deterministic! First: %v, Repeat %d: %v", first, i, repeat)
		}
	}

	nameVal := "Sarah Connor"
	firstName := MaskValue("name", nameVal, StrategyFaker)
	for i := 0; i < 20; i++ {
		repeatName := MaskValue("name", nameVal, StrategyFaker)
		if repeatName != firstName {
			t.Fatalf("Faker name is not deterministic! First: %v, Repeat %d: %v", firstName, i, repeatName)
		}
	}
}

func TestNilEmptySpecialChars(t *testing.T) {
	// Nil input
	if got := MaskValue("email", nil, StrategyPartial); got != nil {
		t.Errorf("expected nil for nil input, got: %v", got)
	}

	// Empty string input
	if got := MaskValue("email", "", StrategyPartial); got != "" {
		t.Errorf("expected empty string for empty input, got: %v", got)
	}

	// Non-PII column should return untouched value
	if got := MaskValue("status", "pending", StrategyPartial); got != "pending" {
		t.Errorf("expected untouched status, got: %v", got)
	}

	// Special chars in name
	specialName := MaskValue("name", "A B C", StrategyPartial).(string)
	if specialName == "" {
		t.Errorf("expected masked special name, got empty")
	}

	// Unicode characters
	unicodeName := MaskValue("name", "José Müller", StrategyPartial).(string)
	if !strings.Contains(unicodeName, "***") {
		t.Errorf("expected masked unicode name, got: %s", unicodeName)
	}
}

func TestMaskRecord(t *testing.T) {
	cols := []string{"id", "email", "name", "price"}
	record := []string{"101", "test@domain.com", "Alice Brown", "45.00"}

	masked := MaskRecord(cols, record, StrategyRedact)
	expected := []string{"101", "[REDACTED]", "[REDACTED]", "45.00"}
	if !reflect.DeepEqual(masked, expected) {
		t.Errorf("MaskRecord mismatch.\nGot:      %v\nExpected: %v", masked, expected)
	}

	maskedPartial := MaskRecord(cols, record, StrategyPartial)
	if maskedPartial[0] != "101" || maskedPartial[3] != "45.00" {
		t.Errorf("non-PII columns modified in MaskRecord: %v", maskedPartial)
	}
	if !strings.Contains(maskedPartial[1], "***@domain.com") {
		t.Errorf("expected partial email mask, got: %s", maskedPartial[1])
	}
}
